// fumadocs-mdx/dist/chunk-KILFIBVW.js
import {
  createDocMethods
} from "./chunk-NVRDCY6Z.js";

// src/runtime/next/index.ts
var _runtime = {
  doc(files) {
    return files.map((file) => {
      const data = file.data;
      return {
        _exports: data,
        body: data.default,
        lastModified: data.lastModified,
        toc: data.toc,
        structuredData: data.structuredData,
        extractedReferences: data.extractedReferences,
        ...data.frontmatter,
        ...createDocMethods(file.info, () => Promise.resolve(data))
      };
    });
  },
  meta(files) {
    return files.map((file) => {
      return {
        info: file.info,
        ...file.data
      };
    });
  },
  docs(docs, metas) {
    const parsedDocs = this.doc(docs);
    const parsedMetas = this.meta(metas);
    return {
      docs: parsedDocs,
      meta: parsedMetas,
      toFumadocsSource() {
        return createMDXSource(parsedDocs, parsedMetas);
      }
    };
  }
};
function createMDXSource(docs, meta = []) {
  return {
    files: resolveFiles({
      docs,
      meta
    })
  };
}
function resolveFiles({ docs, meta }) {
  const outputs = [];
  for (const entry of docs) {
    outputs.push({
      type: "page",
      absolutePath: entry.info.fullPath,
      path: entry.info.path,
      data: entry
    });
  }
  for (const entry of meta) {
    outputs.push({
      type: "meta",
      absolutePath: entry.info.fullPath,
      path: entry.info.path,
      data: entry
    });
  }
  return outputs;
}

export {
  _runtime,
  createMDXSource,
  resolveFiles
};
