// fumadocs-mdx/dist/chunk-4757L6ST.js
import {
  metaLoaderGlob,
  validate
} from "./chunk-2AQRQXSO.js";

// src/loaders/meta.ts
import { dump, load } from "js-yaml";
import { z } from "zod";
var querySchema = z.object({
  collection: z.string().optional()
}).loose();
function createMetaLoader(configLoader, resolve = {}) {
  const { json: resolveJson = "js", yaml: resolveYaml = "js" } = resolve;
  function stringifyOutput(isJson, data) {
    if (isJson) {
      return resolveJson === "json" ? JSON.stringify(data) : `export default ${JSON.stringify(data)}`;
    } else {
      return resolveYaml === "yaml" ? dump(data) : `export default ${JSON.stringify(data)}`;
    }
  }
  return {
    test: metaLoaderGlob,
    async load({ filePath, query, getSource }) {
      const parsed = querySchema.parse(query);
      const collection = parsed.collection ? (await configLoader.getConfig()).getCollection(parsed.collection) : void 0;
      if (!collection) return null;
      const isJson = filePath.endsWith(".json");
      const source = await getSource();
      let data;
      try {
        data = isJson ? JSON.parse(source) : load(source);
      } catch (e) {
        throw new Error(`invalid data in ${filePath}`, { cause: e });
      }
      let schema;
      switch (collection?.type) {
        case "meta":
          schema = collection.schema;
          break;
        case "docs":
          schema = collection.meta.schema;
          break;
      }
      if (schema) {
        data = await validate(
          schema,
          data,
          { path: filePath, source },
          `invalid data in ${filePath}`
        );
      }
      return {
        code: stringifyOutput(isJson, data)
      };
    },
    bun: {
      async fallback({ getSource, filePath }) {
        const source = await getSource();
        const isJson = filePath.endsWith(".json");
        let data;
        try {
          data = isJson ? JSON.parse(source) : load(source);
        } catch (e) {
          throw new Error(`invalid data in ${filePath}`, { cause: e });
        }
        return {
          loader: "object",
          exports: data
        };
      }
    }
  };
}

export {
  createMetaLoader
};
