// fumadocs-mdx/dist/chunk-QXHN25N3.js
import {
  buildConfig
} from "./chunk-2HXTGJBI.js";

// src/config/load-from-file.ts
import * as path from "path";
import { pathToFileURL } from "url";
async function compileConfig(configPath, outDir) {
  const { build } = await import("esbuild");
  const transformed = await build({
    entryPoints: [{ in: configPath, out: "source.config" }],
    bundle: true,
    outdir: outDir,
    target: "node20",
    write: true,
    platform: "node",
    format: "esm",
    packages: "external",
    outExtension: {
      ".js": ".mjs"
    },
    allowOverwrite: true
  });
  if (transformed.errors.length > 0) {
    throw new Error("failed to compile configuration file");
  }
}
async function loadConfig(configPath, outDir, build = false) {
  if (build) await compileConfig(configPath, outDir);
  const url = pathToFileURL(path.resolve(outDir, "source.config.mjs"));
  url.searchParams.set("hash", Date.now().toString());
  const config = import(url.href).then(
    (loaded) => buildConfig(loaded)
  );
  return await config;
}

export {
  loadConfig
};
