// fumadocs-mdx/dist/plugins/json-schema.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/plugins/json-schema.ts
var json_schema_exports = {};
__export(json_schema_exports, {
  default: () => jsonSchema
});
module.exports = __toCommonJS(json_schema_exports);
var import_zod = require("zod");
var import_promises = __toESM(require("fs/promises"), 1);
var import_node_path = __toESM(require("path"), 1);
function jsonSchema({
  insert = false
} = {}) {
  let config;
  function getSchemaPath(name) {
    return `json-schema/${name}.json`;
  }
  return {
    config(v) {
      config = v;
    },
    configureServer(server) {
      if (!server.watcher || !insert) return;
      server.watcher.on("add", async (file) => {
        let parent;
        let match;
        for (const collection of config.collectionList) {
          if (collection.type === "meta" && collection.hasFile(file)) {
            match = collection;
            break;
          }
          if (collection.type === "docs" && collection.meta.hasFile(file)) {
            parent = collection;
            match = collection.meta;
            break;
          }
        }
        if (!match) return;
        let obj;
        try {
          const content = (await import_promises.default.readFile(file)).toString();
          obj = content.length > 0 ? JSON.parse(content) : {};
        } catch {
          return;
        }
        if ("$schema" in obj) return;
        const schemaPath = import_node_path.default.join(
          this.outDir,
          getSchemaPath(parent ? `${parent.name}.meta` : match.name)
        );
        const updated = {
          $schema: import_node_path.default.relative(import_node_path.default.dirname(file), schemaPath),
          ...obj
        };
        await import_promises.default.writeFile(file, JSON.stringify(updated, null, 2));
      });
    },
    emit() {
      const files = [];
      function onSchema(name, schema) {
        files.push({
          path: getSchemaPath(name),
          content: JSON.stringify(
            import_zod.z.toJSONSchema(schema, {
              io: "input",
              unrepresentable: "any"
            })
          )
        });
      }
      for (const collection of config.collectionList) {
        if (collection.type === "docs") {
          if (collection.meta.schema instanceof import_zod.z.ZodType) {
            onSchema(`${collection.name}.meta`, collection.meta.schema);
          }
          if (collection.docs.schema instanceof import_zod.z.ZodType) {
            onSchema(`${collection.name}.docs`, collection.docs.schema);
          }
        } else if (collection.schema instanceof import_zod.z.ZodType) {
          onSchema(collection.name, collection.schema);
        }
      }
      return files;
    }
  };
}
