// fumadocs-ui/dist/components/callout.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { CircleCheck, CircleX, Info, TriangleAlert } from '../icons.js';
import { cn } from '../utils/cn.js';
const iconClass = 'size-5 -me-0.5 fill-(--callout-color) text-fd-card';
export function Callout({ children, title, ...props }) {
    return (_jsxs(CalloutContainer, { ...props, children: [title && _jsx(CalloutTitle, { children: title }), _jsx(CalloutDescription, { children: children })] }));
}
function resolveAlias(type) {
    if (type === 'warn')
        return 'warning';
    if (type === 'tip')
        return 'info';
    return type;
}
export function CalloutContainer({ type: inputType = 'info', icon, children, className, style, ...props }) {
    const type = resolveAlias(inputType);
    return (_jsxs("div", { className: cn('flex gap-2 my-4 rounded-xl border bg-fd-card p-3 ps-1 text-sm text-fd-card-foreground shadow-md', className), style: {
            '--callout-color': `var(--color-fd-${type}, var(--color-fd-muted))`,
            ...style,
        }, ...props, children: [_jsx("div", { role: "none", className: "w-0.5 bg-(--callout-color)/50 rounded-sm" }), icon ??
                {
                    info: _jsx(Info, { className: iconClass }),
                    warning: _jsx(TriangleAlert, { className: iconClass }),
                    error: _jsx(CircleX, { className: iconClass }),
                    success: _jsx(CircleCheck, { className: iconClass }),
                }[type], _jsx("div", { className: "flex flex-col gap-2 min-w-0 flex-1", children: children })] }));
}
export function CalloutTitle({ children, className, ...props }) {
    return (_jsx("p", { className: cn('font-medium my-0!', className), ...props, children: children }));
}
export function CalloutDescription({ children, className, ...props }) {
    return (_jsx("div", { className: cn('text-fd-muted-foreground prose-no-margin empty:hidden', className), ...props, children: children }));
}
