// fumadocs-ui/dist/components/dynamic-codeblock.js
'use client';
import { jsx as _jsx } from "react/jsx-runtime";
import { CodeBlock, Pre } from '../components/codeblock.js';
import { useShiki } from 'fumadocs-core/highlight/client';
import { cn } from '../utils/cn.js';
import { createContext, Suspense, use, useDeferredValue, useId, } from 'react';
const PropsContext = createContext(undefined);
function DefaultPre(props) {
    const extraProps = use(PropsContext);
    return (_jsx(CodeBlock, { ...props, ...extraProps, className: cn('my-0', props.className, extraProps?.className), children: _jsx(Pre, { children: props.children }) }));
}
export function DynamicCodeBlock({ lang, code, codeblock, options, wrapInSuspense = true, }) {
    const id = useId();
    const shikiOptions = {
        lang,
        ...options,
        components: {
            pre: DefaultPre,
            ...options?.components,
        },
    };
    const children = (_jsx(PropsContext, { value: codeblock, children: _jsx(Internal, { id: id, ...useDeferredValue({ code, options: shikiOptions }) }) }));
    if (wrapInSuspense)
        return (_jsx(Suspense, { fallback: _jsx(Placeholder, { code: code, components: shikiOptions.components }), children: children }));
    return children;
}
function Placeholder({ code, components = {}, }) {
    const { pre: Pre = 'pre', code: Code = 'code' } = components;
    return (_jsx(Pre, { children: _jsx(Code, { children: code.split('\n').map((line, i) => (_jsx("span", { className: "line", children: line }, i))) }) }));
}
function Internal({ id, code, options, }) {
    return useShiki(code, options, [id, options.lang, code]);
}
