// fumadocs-core/dist/chunk-ADBHPKXG.js
// src/mdx-plugins/remark-mdx-files.ts
import { visit } from "unist-util-visit";
function parseFileTree(code) {
  const lines = code.split(/\r?\n/);
  const stack = /* @__PURE__ */ new Map();
  for (const line of lines) {
    let depth = 0;
    let name = line;
    let match;
    while (match = /(?:├──|│|└──)\s*/.exec(name)) {
      name = name.slice(match[0].length);
      depth++;
    }
    if (!name) continue;
    const node = name.endsWith("/") ? { type: "folder", name, children: [], depth } : { type: "file", name, depth };
    let parent;
    for (let i = depth - 1; i >= 0 && !parent; i--) {
      parent = stack.get(i);
    }
    stack.set(depth, node);
    if (!parent) continue;
    if (parent.type === "file") {
      Object.assign(parent, {
        type: "folder",
        children: []
      });
    }
    parent.children.push(node);
  }
  return stack.get(0);
}
function defaultToMDX(node, depth = 0) {
  if (depth === 0) {
    return {
      type: "mdxJsxFlowElement",
      name: "Files",
      attributes: [],
      children: [defaultToMDX(node, depth + 1)]
    };
  }
  const attributes = [
    { type: "mdxJsxAttribute", name: "name", value: node.name }
  ];
  if (node.type === "file") {
    return {
      type: "mdxJsxFlowElement",
      attributes,
      children: [],
      name: "File"
    };
  }
  attributes.push({
    type: "mdxJsxAttribute",
    name: "defaultOpen",
    value: null
  });
  return {
    type: "mdxJsxFlowElement",
    attributes,
    name: "Folder",
    children: node.children.map((item) => defaultToMDX(item, depth + 1))
  };
}
function remarkMdxFiles(options = {}) {
  const { lang = "files", toMdx = defaultToMDX } = options;
  return (tree) => {
    visit(tree, "code", (node) => {
      if (node.lang !== lang || !node.value) return;
      const fileTree = parseFileTree(node.value);
      if (!fileTree) return;
      Object.assign(node, toMdx(fileTree));
    });
  };
}

export {
  remarkMdxFiles
};
