// fumadocs-core/dist/chunk-A4G5V4FQ.js
// src/mdx-plugins/mdast-utils.ts
import { valueToEstree } from "estree-util-value-to-estree";
function flattenNode(node) {
  if ("children" in node)
    return node.children.map((child) => flattenNode(child)).join("");
  if ("value" in node) return node.value;
  return "";
}
function toMdxExport(name, value) {
  return {
    type: "mdxjsEsm",
    value: "",
    data: {
      estree: {
        type: "Program",
        sourceType: "module",
        body: [
          {
            type: "ExportNamedDeclaration",
            attributes: [],
            specifiers: [],
            source: null,
            declaration: {
              type: "VariableDeclaration",
              kind: "let",
              declarations: [
                {
                  type: "VariableDeclarator",
                  id: {
                    type: "Identifier",
                    name
                  },
                  init: valueToEstree(value)
                }
              ]
            }
          }
        ]
      }
    }
  };
}

export {
  flattenNode,
  toMdxExport
};
