// fumadocs-core/dist/chunk-GLRQBLGN.js
import {
  generateCodeBlockTabs
} from "./chunk-TWIDBWFG.js";

// src/mdx-plugins/remark-npm.ts
import { visit } from "unist-util-visit";
import convert from "npm-to-yarn";
var aliases = ["npm", "package-install"];
function remarkNpm({
  persist = false,
  packageManagers = [
    { command: (cmd) => convert(cmd, "npm"), name: "npm" },
    { command: (cmd) => convert(cmd, "pnpm"), name: "pnpm" },
    { command: (cmd) => convert(cmd, "yarn"), name: "yarn" },
    { command: (cmd) => convert(cmd, "bun"), name: "bun" }
  ]
} = {}) {
  return (tree) => {
    visit(tree, "code", (node) => {
      if (!node.lang || !aliases.includes(node.lang)) return;
      let code = node.value;
      if (node.lang === "package-install" && !code.startsWith("npm") && !code.startsWith("npx")) {
        code = `npm install ${code}`;
      }
      const options = {
        persist,
        tabs: [],
        triggers: []
      };
      for (const manager of packageManagers) {
        const value = manager.value ?? manager.name;
        const command = manager.command(code);
        if (!command || command.length === 0) continue;
        options.defaultValue ??= value;
        options.triggers.push({
          value,
          children: [{ type: "text", value: manager.name }]
        });
        options.tabs.push({
          value,
          children: [
            {
              type: "code",
              lang: "bash",
              meta: node.meta,
              value: command
            }
          ]
        });
      }
      Object.assign(node, generateCodeBlockTabs(options));
      return "skip";
    });
  };
}

export {
  remarkNpm
};
